File: pclassic/2016s/PClassic2016Stubs/stubs/BlackMarket.java
import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;

public class BlackMarket {
  //Fill out the body of this method
  public static boolean freeMoney(double[][] prices){
    return false;
  }
	
  public static void main(String[] args) throws NumberFormatException, IOException {
    BufferedReader b = new BufferedReader(new FileReader("BlackMarketIN.txt"));

    while(b.ready()){
      String in = b.readLine();
      String[] inArr = in.split(",");
      double[][] prices = new double[inArr.length][inArr.length];
			
      for (int i = 0; i < inArr.length; i++){
        String[] row = inArr[i].split(" ");
        for(int j = 0; j < inArr.length; j++){
          prices[i][j] = Double.parseDouble(row[j].trim());
        }
      }
      System.out.println(freeMoney(prices));

    }
		
  }

}

File: pclassic/2016s/PClassic2016Stubs/stubs/BlackMarket.py
import math

def can_profit(prices):
    N = len(prices)
    logprices = []
    dist = []
    for row in prices:
        logprices.append([math.log(x) for x in row])
        dist.append(1e400)
    dist[0] = 0.0
    for i in range(N-1):
        for j in range(N):
            for k in range(N):
                dist[k] = min(dist[k], dist[j] + logprices[j][k])
    for j in range(N):
        for k in range(N):
            if dist[k] > dist[j] + logprices[j][k]:
                return True
    return False

if __name__ == "__main__":
    with open("BlackMarketIN.txt", "r") as f:
        while True:
            s = f.readline()
            if s == "":
                break
            rows = s.split(",")
            prices = []
            for row in rows:
                prices.append([float(x) for x in row.split(" ")])
            print("true" if can_profit(prices) else "false")
